### ISSUE TITLE
The same seed gives different results

### ISSUE THREAD

**Issue by nagadomi** [soumith/cudnn.torch#92]
I am using torch.manualSeed() and cutorch.manualSeed() in my code. However, I get different result every time when using cudnn module. nn module works normally.

Reproducing:
``` lua
require 'cutorch'
require 'cunn'
require 'cudnn'
require 'optim'

local SpatialConvolution = cudnn.SpatialConvolution
local SpatialMaxPooling = cudnn.SpatialMaxPooling
--local SpatialConvolution = nn.SpatialConvolution
--local SpatialMaxPooling = nn.SpatialMaxPooling

cudnn.benchmark = false
torch.setdefaulttensortype("torch.FloatTensor")
torch.manualSeed(71)
cutorch.manualSeed(71)

local model = nn.Sequential()
model:add(nn.View(30, 48, 48))
model:add(SpatialConvolution(30, 64, 3, 3, 1, 1, 1, 1))
model:add(nn.ReLU())
model:add(SpatialConvolution(64, 64, 3, 3, 1, 1, 1, 1))
model:add(nn.ReLU())
model:add(SpatialMaxPooling(2, 2, 2, 2))
model:add(SpatialConvolution(64, 128, 3, 3, 1, 1, 1, 1))
model:add(nn.ReLU())
model:add(SpatialConvolution(128, 128, 3, 3, 1, 1, 1, 1))
model:add(nn.ReLU())
model:add(SpatialMaxPooling(2, 2, 2, 2))
model:add(nn.View(128 * 12 * 12))
model:add(nn.Linear(128 * 12 * 12, 512))
model:add(nn.Dropout(0.5))
model:add(nn.Linear(512, 1))

local inputs = torch.Tensor(32, 1, 30, 48, 48):uniform():cuda()
local targets = torch.Tensor(32, 1):uniform():cuda()
local criterion = nn.MSECriterion():cuda()
local config = { learningRate = 0.00001 }

model = model:cuda()
model:training()
local parameters, gradParameters = model:getParameters()
for i = 1, 100 do
   local feval = function(x)
      if x ~= parameters then
	 parameters:copy(x)
      end
      gradParameters:zero()
      inputs = inputs:clone()
      local output = model:forward(inputs)
      local f = criterion:forward(output, targets)
      model:backward(inputs, criterion:backward(output, targets))
      return f, gradParameters
   end
   optim.adam(feval, parameters, config)
   if i % 10 == 0 then
      collectgarbage()
   end
end
model:evaluate()
local y = model:forward(inputs)
print(y:sum())

--[[
# on GTX760, CUDA 7.5, cuDNN v3

# cunn - same result

% th seed.lua
   18.265419006348
% th seed.lua
   18.265419006348
% th seed.lua
   18.265419006348
% th seed.lua
   18.265419006348

# cudnn - different result

% th seed.lua
   18.265224456787
% th seed.lua
   18.26549911499
% th seed.lua
   18.265830993652
% th seed.lua
   18.266021728516

--]]
```

**Comment by soumith** [comment]
this is expected. cudnn convolutions and pooling are not deterministic implementations.

**Comment by nagadomi** [comment]
Thank you for your explaining.
This means we cannot reproduce the submitted solution in the data prediction competition like a kaggle. :sob:

**Comment by mys007** [comment]
It should be also possible to choose a deterministic algorithm using `SpatialConvolution:setMode()` with proper arguments (see cudnn manual), but the last time I tried it it was slower than the `nn` equivalent for my tensor sizes. But it's worth a try and maybe with cudnn v4 it's even better now. SpatialMaxPoolings from nn and cudnn should be more or less interchangeable in performance, I'd say.

**Comment by nagadomi** [comment]
Thank. I fixed this issue.
```
function SpatialConvolution(nInputPlane, nOutputPlane,
                             kW, kH, dW, dH, padW, padH, groups)
   local conv = cudnn.SpatialConvolution(nInputPlane, nOutputPlane,          
                                          kW, kH, dW, dH, padW, padH, groups)
  -- set deterministic algorithm
   conv:setMode('CUDNN_CONVOLUTION_FWD_ALGO_GEMM',
                'CUDNN_CONVOLUTION_BWD_DATA_ALGO_1',
                'CUDNN_CONVOLUTION_BWD_FILTER_ALGO_1'):fastest()
   return conv
end
```
The performance is close to default algorithm in my setting.

**Comment by soumith** [comment]
After setting mode, i don't think you should do :fastest()
